### ISSUE TITLE
`log4j-bom` leaks non-Log4j dependencies

### ISSUE THREAD

**Issue by wilkinsona** [apache/logging-log4j2#3066]
## Description

`log4j-bom` inherits from `logging-parent`. Unfortunately, this results in `log4j-bom` managing a number of dependencies that are unrelated to a consumer's use of Log4j2. Those dependencies are:

- `biz.aQute.bnd:biz.aQute.bnd.annotation:7.0.0`
- `com.github.spotbugs:spotbugs-annotations:4.8.6`
- `org.jspecify:jspecify:1.0.0`
- `org.osgi:osgi.annotation:8.1.0`
- `org.osgi:org.osgi.annotation.bundle:2.0.0`
- `org.osgi:org.osgi.annotation.versioning:1.1.2`

`org.apache.maven.plugin-tools:maven-plugin-annotations:3.13.1` is also being managed but this is inherited from the `org.apache:apache` pom and has already been [reported](https://github.com/apache/maven-apache-parent/issues/265) and, pending an upgrade to use the new parent, [fixed](https://github.com/apache/maven-apache-parent/pull/266).

This unwanted dependency management can conflict with a user's own dependency management for those dependencies. Depending on how that dependency management is configured, it may override it leaving a consumer using an unexpected version of a dependency.

## Configuration

**Version:** 2.21.0 and later. The list of dependencies above is from 2.24.1.

**Operating system:** Any

**JDK:** Any

## Logs

N/A

## Reproduction

Run `mvn help:effective-pom` in a project with the following `pom.xml`:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<groupId>com.example</groupId>
	<artifactId>log4j-bom-problem</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	<dependencyManagement>
		<dependencies>
			<dependency>
				<groupId>org.apache.logging.log4j</groupId>
				<artifactId>log4j-bom</artifactId>
				<version>2.24.1</version>
				<scope>import</scope>
				<type>pom</type>
			</dependency>
		</dependencies>
	</dependencyManagement>
</project>
```

The output will show a number of managed dependencies that aren't in the `org.apache.logging.log4j` group.

**Comment by vy** [comment]
@wilkinsona, thanks so much for the report! Managed dependencies that don't belong to the `o.a.l.log4j` group are indeed not wanted and should be removed. I will see what we can do about it.

Log4j inherits the BOM flattening logic from `org.apache.logging:logging-parent`, which

1. Inherits from `org.apache:apache`
2. Used by all Maven-based projects of Logging Services
3. [Intentionally keeps the parent!](https://github.com/apache/logging-parent/blob/main/pom.xml#L485)

Quoting from the above linked `pom.xml` snippet:

> Consider the following problem experienced in `log4j-transform-maven-plugin`:
>
> 1. `log4j-transform-parent` depends on `log4j-transform-bom`
> 2. `log4j-transform-bom` depends on `logging-parent`
> 3. `logging-parent` contains `dependencyManagement`, etc. that are used by `log4j-transform-maven-plugin`
> 4. Dependencies of `log4j-transform-maven-plugin` et al. is resolved *at runtime*
> 5. Though at runtime, the deployed `log4j-transform-bom` is used, which is flattened and hence doesn't have a parent!
> 6. Hence, at runtime, all `logging-parent` logic is lost
>
> To avoid this, parents should better be kept while flattening BOMs.

In short, we decided to keep the parent since it was necessary for modules needing dependency resolution at runtime. We need to do some research on what would be the best way to approach this problem. @wilkinsona, your feedback is more than welcome.

**Comment by ppkarwasz** [comment]
@vy,

Should we close this one, since `logging-parent` version `12.0.0` has been released?

**Comment by vy** [comment]
Validated on 07590bc087204ccd4d091c65d0c5da255c92df68 that the installed `~/.m2/repository/org/apache/logging/log4j/log4j-bom/2.25.0-SNAPSHOT/log4j-bom-2.25.0-SNAPSHOT.pom`

1. Contains no `parent`
2. Contains no `properties`
3. Contains only `log4j-*` dependencies in `dependencyManagement`

**Comment by ppkarwasz** [comment]
@dsvensson,

> New content added to gradle verification metadata (nevermind the _"key couldn't be downloaded"_, have keyservers disabled and haven't imported the keys yet):
>
> ...
>
> Is this intended going forward, or wait for patch release?

Yes, the presence of those libraries in the Gradle verification metadata is **expected** — depending on how you consume Log4j.

### Why These Dependencies Appear

The annotation libraries you listed (BND, ErrorProne, JSpecify, SpotBugs, and various OSGi annotations) are **intentionally used by Log4j**, but only useful at **compile time**. To handle this across different build tools:

* In **Maven**, they are declared with the `provided` scope, which is not inherited by downstream projects.
* However, as explained in #3110, this setup can lead to unwanted compiler warnings for some users.
* To address that, since version `2.25.0`, Log4j also publishes Gradle `.module` metadata where these dependencies are marked as `compileOnlyApi`:

  * `compileOnly` means they are **not present at runtime**.
  * `api` means they are **inherited by consumers**.
  
  See also #3450, which implements this change.

### What You Should Do

* If your application uses the [Log4j API](https://logging.apache.org/log4j/2.x/manual/api.html), these annotation libraries might be useful during compilation and will appear in Gradle’s dependency verification..
* If you **do not** use the Log4j API (e.g., you're using SLF4J as your logging API), then you don't need those compile-time dependencies at all.

  In that case, the cleanest setup is to use `log4j-core` and related modules with the `runtimeOnly` scope, as recommended in our [Installation Guide](https://logging.apache.org/log4j/2.x/manual/installation.html):

  ```groovy
  dependencies {
    runtimeOnly platform('org.apache.logging.log4j:log4j-bom:2.25.0')

    // Logging API
    implementation 'org.slf4j:slf4j-api:2.0.17'

    // Logging implementation
    runtimeOnly 'org.apache.logging.log4j:log4j-slf4j2-impl'
    runtimeOnly 'org.apache.logging.log4j:log4j-core'
  }
  ```

  This setup avoids including unnecessary compile-time dependencies and keeps your verification metadata clean.
